### ISSUE TITLE
Cannot find Powershell when building with VS 2019 and manifest file.

### ISSUE THREAD

**Issue by j-horner-c4x** [microsoft/vcpkg#21203]
**Describe the bug**
Building with Visual Studio with `vcpkg` as a submodule and manifest `vcpkg.json` file fails with the error:
```
'C:\Windows\System32\powershell' is not recognized as an internal or external command
```

**Environment**
- OS: Windows 10 Pro, Version 20H2, OS build 19042.1288, Windows Feature Experience Pack 120.2212.3920.0
- Compiler: Visual Studio 2019 Professional 16.10.4
- Installed components: 
    - Desktop development with C++
    - Linux development with C++
    - C++ Clang-cl for v142 build tools (x64/x86)
    - C++ Clang Compiler for Windows (11.0.0)
- `vcpkg` version tag 2021.05.12 

**To Reproduce**
1. Clone this repo (and its vcpkg submodule): https://github.com/j-horner-c4x/vs-vcpkg-bug
    - It is a minimal VS generated CMake project with `CMakePresets.json` and `vcpkg.json`
2. Open folder with Visual Studio
3. Generate cache
4. Build fails with error

**Expected behavior**
CMake cache generation to succeed and project to compile.

**Failure logs** 
[build_fail_output.txt](https://github.com/microsoft/vcpkg/files/7485950/build_fail_output.txt)
- Summary of attached:
```
1> [CMake] -- Running vcpkg install - done
1> [CMake] -- The C compiler identification is MSVC 19.29.30040.0
1> [CMake] -- The CXX compiler identification is MSVC 19.29.30040.0
1> [CMake] -- Detecting C compiler ABI info
1> [CMake] -- Detecting C compiler ABI info - failed
1> [CMake] -- Check for working C compiler: C:/Program Files (x86)/Microsoft Visual Studio/2019/Professional/VC/Tools/MSVC/14.29.30037/bin/Hostx64/x64/cl.exe
1> [CMake] -- Check for working C compiler: C:/Program Files (x86)/Microsoft Visual Studio/2019/Professional/VC/Tools/MSVC/14.29.30037/bin/Hostx64/x64/cl.exe - broken
1> [CMake] CMake Error at C:/Program Files (x86)/Microsoft Visual Studio/2019/Professional/Common7/IDE/CommonExtensions/Microsoft/CMake/CMake/share/cmake-3.20/Modules/CMakeTestCCompiler.cmake:66 (message):
1> [CMake]   The C compiler
1> [CMake] 
1> [CMake]     "C:/Program Files (x86)/Microsoft Visual Studio/2019/Professional/VC/Tools/MSVC/14.29.30037/bin/Hostx64/x64/cl.exe"
1> [CMake] 
1> [CMake]   is not able to compile a simple test program.
1> [CMake] 
1> [CMake]   It fails with the following output:
1> [CMake] 
1> [CMake]     Change Dir: C:/Users/jhorner/source/repos/CMakeProject1/build/test-build/CMakeFiles/CMakeTmp
1> [CMake]     
1> [CMake]     Run Build Command(s):C:/Program Files (x86)/Microsoft Visual Studio/2019/Professional/Common7/IDE/CommonExtensions/Microsoft/CMake/Ninja/ninja.exe cmTC_a390e && [1/2] Building C object CMakeFiles\cmTC_a390e.dir\testCCompiler.c.obj
1> [CMake]     [2/2] Linking C executable cmTC_a390e.exe
1> [CMake]     FAILED: cmTC_a390e.exe 
1> [CMake]     cmd.exe /C "cd . && "C:\Program Files (x86)\Microsoft Visual Studio\2019\Professional\Common7\IDE\CommonExtensions\Microsoft\CMake\CMake\bin\cmake.exe" -E vs_link_exe --intdir=CMakeFiles\cmTC_a390e.dir --rc=C:\PROGRA~2\WI3CF2~1\10\bin\100190~1.0\x64\rc.exe --mt=C:\PROGRA~2\WI3CF2~1\10\bin\100190~1.0\x64\mt.exe --manifests  -- C:\PROGRA~2\MICROS~4\2019\PROFES~1\VC\Tools\MSVC\1429~1.300\bin\Hostx64\x64\link.exe /nologo CMakeFiles\cmTC_a390e.dir\testCCompiler.c.obj  /out:cmTC_a390e.exe /implib:cmTC_a390e.lib /pdb:cmTC_a390e.pdb /version:0.0 /machine:x64  /debug /INCREMENTAL /subsystem:console  kernel32.lib user32.lib gdi32.lib winspool.lib shell32.lib ole32.lib oleaut32.lib uuid.lib comdlg32.lib advapi32.lib && cmd.exe /C "cd /D C:\Users\jhorner\source\repos\CMakeProject1\build\test-build\CMakeFiles\CMakeTmp && C:\Windows\System32\powershell -noprofile -executionpolicy Bypass -file C:/Users/jhorner/source/repos/CMakeProject1/external/vcpkg/scripts/buildsystems/msbuild/applocal.ps1 -targetBinary C:/Users/jhorner/source/repos/CMakeProject1/build/test-build/CMakeFiles/CMakeTmp/cmTC_a390e.exe -installedDir C:/Users/jhorner/source/repos/CMakeProject1/external/vcpkg/installed/x64-windows/debug/bin -OutVariable out""
1> [CMake]     'C:\Windows\System32\powershell' is not recognized as an internal or external command,
1> [CMake]     operable program or batch file.
1> [CMake]     ninja: build stopped: subcommand failed.
``` 

**Additional context**
Looks similiar to: https://github.com/microsoft/vcpkg/issues/17448
This works on my home desktop but not on my work computer so its possibly a misconfigured environment. Any suggested work around would be appreciated.

**Comment by j-horner-c4x** [comment]
Doing the following did not fix the issue:

- Removing `vcpkg.json`
- Manually running `bootstrap-vcpkg.bat -disableMetrics` on a separate `vcpkg` copy and using that instead.

So far the only fix I've found is to remove the following from `CMakePresets.json`:
```
          "CMAKE_TOOLCHAIN_FILE": {
            "value": "${sourceDir}/external/vcpkg/scripts/buildsystems/vcpkg.cmake",
            "type": "FILEPATH"
          },
```

**Comment by j-horner-c4x** [comment]
Setting `"generator": "Visual Studio 16 2019"` and `"architecture.strategy": "set"` also fails:
[build_fail_output.txt](https://github.com/microsoft/vcpkg/files/7496389/build_fail_output.txt)

**Comment by JonLiu1993** [comment]
@j-horner-c4x ,Thanks posting this issue. I can't reproduce your problem locally based on your steps. Did I miss any step?
````
 [CMake] CMake Error at C:/Program Files (x86)/Microsoft Visual Studio/2019/Enterprise/Common7/IDE/CommonExtensions/Microsoft/CMake/CMake/share/cmake-3.20/Modules/CMakeDetermineSystem.cmake:129 (message):
1> [CMake]   Could not find toolchain file:
1> [CMake]   F:/test-pr/test/vs-vcpkg-bug/external/vcpkg/scripts/buildsystems/vcpkg.cmake
1> [CMake] Call Stack (most recent call first):
1> [CMake]   CMakeLists.txt:6 (project)
1> [CMake] CMake Error: CMAKE_C_COMPILER not set, after EnableLanguage
1> [CMake] CMake Error: CMAKE_CXX_COMPILER not set, after EnableLanguage
1> [CMake] -- Configuring incomplete, errors occurred!
````

**Comment by j-horner-c4x** [comment]
Thanks for looking into this. It can't find the toolchain file so I think you might not have cloned the `vcpkg` submodule.
Try:
```
git submodule init
git submodule update
```
Sorry I should've been more explicit with that.

Also I think VS should select the `test-build` preset and set the C++ compiler automatically?
You might need to set the option to enable CMake Presets, described here: https://docs.microsoft.com/en-us/cpp/build/cmake-presets-vs?view=msvc-160#-enable--cmakepresetsjson-integration-in-visual-studio-2019
But I think they changed it in a recent version? That bit just worked automatically for me.

**Comment by JonLiu1993** [comment]
@j-horner-c4x ,Thanks for your help, I will reproduce your issue locally , thanks again

**Comment by j-horner-c4x** [comment]
Also doesn't work with version 16.11.6. I also submitted a ticket to Visual Studio support which included some more diagnostics file. I've also added the build output files generated by CMake below.
[build_output.zip](https://github.com/microsoft/vcpkg/files/7514614/build_output.zip)
s.

**Comment by BillyONeal** [comment]
> **Additional context**
> Looks similiar to: #17448

Probably not, considering the copy of powershell attempted to be used here is from system32, not from the store.

The funny powershell path is probably coming from `find_program(Z_VCPKG_BUILTIN_POWERSHELL_PATH powershell REQUIRED) ` in:

https://github.com/microsoft/vcpkg/blob/de88cc55c4acd0cdd955e5bd4c412e1b9250f29e/scripts/buildsystems/vcpkg.cmake#L175-L198

Do you have a `C:\Windows\System32\powershell.exe` on the problematic system? Is `exe` in your `PATHEXT` ?

**Comment by j-horner-c4x** [comment]
It's odd. That file does exist on the problematic system but it's an empty file and has no extension. On the work system, that file doesn't exist (with or without the extension).

When I launch Powershell I believe its running this file:
`%SystemRoot%\system32\WindowsPowerShell\v1.0\powershell.exe`

My `PATHEXT`: `.COM;.EXE;.BAT;.CMD;.VBS;.VBE;.JS;.JSE;.WSF;.WSH;.MSC`

**Comment by BillyONeal** [comment]
> It's odd. That file does exist on the problematic system but it's an empty file and has no extension. On the working system, that file doesn't exist (with or without the extension).
> 

Hmmm that's sufficiently odd that I don't think we're going to look to add defenses for that unless we know the specific condition is common. I do note, however, that we do want to get powershell out of this business entirely, which would resolve the problem.

**Comment by j-horner-c4x** [comment]
Removing the file `C:\Windows\System32\powershell` which was incorrectly placed on the system resolved the issue.

Thank you.

**Comment by BillyONeal** [comment]
OK, I'm going to mark this one closed but leave #17448 open as the general "powershell continues to be a problematic dependency" issue. Thanks for confirming!
